### PR TITLE
Fix drop behavior #11

### PR DESCRIPTION
**Pull request by tajtiattila**
This change should fix the `Drop` issue in #11.

There is also a new test drop_keeps_strong_refs to verify the expected behavior.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -509,23 +509,31 @@ impl<T> NodeData<T> {
 
 impl<T> Drop for NodeData<T> {
     fn drop(&mut self) {
-        // Collect all descendant nodes and detach them to prevent the stack overflow.
```
Comment (RazrFalcon) [inline src/lib.rs:512]: I think you should keep this comment. Otherwise it's not clear why we need a custom Drop.

File: src/lib.rs [hunk 2]
```diff
@@ -509,23 +509,31 @@ impl<T> NodeData<T> {
...
-            }
+        if let Some(child) = self.first_child.take() {
+            detach_all_unique(child);
         }
+    }
+}
+
+/// Detach all unique descendant and sibling nodes recursively.
+fn detach_all_unique<T>(node: Link<T>) {
+    let mut work = vec![node];
```
Comment (RazrFalcon) [inline src/lib.rs]: `work` is a weird name. Maybe we should use something like `remaining`?

**Review by RazrFalcon — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 3]
```diff
@@ -509,23 +509,31 @@ impl<T> NodeData<T> {
...
-                }
-            }
+        if let Some(child) = self.first_child.take() {
+            detach_all_unique(child);
         }
+    }
+}
+
+/// Detach all unique descendant and sibling nodes recursively.
+fn detach_all_unique<T>(node: Link<T>) {
```
Comment (RazrFalcon) [inline src/lib.rs]: Do we need a separate method? We can keep it inside Drop.

**Review by RazrFalcon — CHANGES_REQUESTED** [review]

**Comment by RazrFalcon** [comment]
So if I understood correctly, the main change here is that only uniquely owned nodes are dropped now, right? There are no other changes.

File: src/lib.rs (hunk 1 above)
Comment (tajtiattila) [reply src/lib.rs:512]: Agreed.

File: src/lib.rs (hunk 2 above)
Comment (tajtiattila) [reply src/lib.rs]: It holds nodes to be processed, but I agree we should try to find a better name. Maybe `open` or `open_set` as in the set of items to check? The Wikipedia page for the A* search algorithm uses that. The old method used `stack` that could also work here.

File: src/lib.rs (hunk 3 above)
Comment (tajtiattila) [reply src/lib.rs]: Yep, not needed. I was still thinking if detach_all_unique needs to be called for the `next_sibling` at the time of writing.

**Review by RazrFalcon — APPROVED** [review]

**Pull request merged**